spec/rspec/rails/matchers/have_reported_error_spec.rb

```diff
@@ -0,0 +1,233 @@
+    it "passes when attributes match exactly" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"), user_id: 123, context: "test")
+      end
+
+      expect(test_block).to have_reported_error.with(user_id: 123, context: "test")
+    end
+
+    it "passes with partial attribute matching" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"), user_id: 123, context: "test", extra: "data")
+      end
+
+      expect(test_block).to have_reported_error.with(user_id: 123)
+    end
+
+    it "passes with hash matching using RSpec matchers" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"), params: { foo: "bar", baz: "qux" })
+      end
+
+      expect(test_block).to have_reported_error.with(params: a_hash_including(foo: "bar"))
+    end
+
+    it "fails when attributes do not match" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"), user_id: 123, context: "actual")
+      end
+      matcher = have_reported_error.with(user_id: 456, context: "expected")
+
+      expect(matcher.matches?(test_block)).to be false
+    end
+
+    it "fails when no error is reported but attributes are expected" do
+      test_block = proc { "no error" }
+      matcher = have_reported_error.with(user_id: 123)
+
+      expect(matcher.matches?(test_block)).to be false
+    end
+  end
+
+  describe "cleanup behavior" do
+    it "unsubscribes from error reporter on successful completion" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"))
+      end
+
+      expect(mock_error_reporter).to receive(:unsubscribe)
+
+      expect(test_block).to have_reported_error
+    end
+
+    it "unsubscribes from error reporter even when exception is raised" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("test"))
+        raise "unexpected error"
+      end
+
+      expect(mock_error_reporter).to receive(:unsubscribe)
+
+      expect {
+        have_reported_error.matches?(test_block)
+      }.to raise_error("unexpected error")
+    end
+  end
+
+  describe "block expectations support" do
+    it "declares support for block expectations" do
+      matcher = have_reported_error
+      expect(matcher).to respond_to(:supports_block_expectations?)
+      expect(matcher.supports_block_expectations?).to be true
+    end
+  end
+
+  describe "integration with actual usage patterns" do
+    it "works with multiple error reports in a block" do
+      test_block = proc do
+        Rails.error.report(StandardError.new("first error"))
+        Rails.error.report(TestError.new("second error"))
+      end
+
+      expect(test_block).to have_reported_error(StandardError)
+    end
+
+    it "works with matcher chaining" do
+      test_block = proc do
+        Rails.error.report(TestError.new("test"), user_id: 123)
+      end
+
+      expect(test_block).to have_reported_error(TestError).and have_reported_error
```

Comment:
Does it really pass when two errors were reported and we expect to have those two reported? Does the matcher fail if either of the errors was not reported?

Reply:
I'm not quite sure how to handle this case. Should this not allow chaining?

Reply:
I’m on the fence with this. It might be useful, but is it worth the effort in this introductory PR?
The matcher _can_ be made chainable, but is it worth the effort?

Reply:
I disabled "and" chaining, but not sure how canonical this method is to rspec-rails :)

Reply:
It’s a bummer that [include](https://github.com/rspec/rspec-rails/blob/6135d5d0bdfd80d6e3051428639facbab6bd831f/lib/rspec/rails/matchers/base_matcher.rb#L8) Composable is in the base class.